### PR TITLE
[NUI] Refactoring key focusable feature for supporting non-focusable browsing.

### PR DESCRIPTION
**Pull request by everLEEst**
We have 3 different cases in scrollablebase on focusable browsing.

1. All children can be focusable.
2. sparsely focusable.
3. none of children are focusable.

in any cases,
user can be browsable all content via scroll and can miss any readable elements.
To do this,
firstly, scrollableBase should be focusable.

and when next focusable child is not in the visible area,
we will move step distance.

we add one property for this.
StepScrollDistance
StepScrollDistanceProperty
which user can modify the distance of this step movement.
but if user didn't set any value, scroll will move only quater size of scroll-axis length.

we added samples for above 3 cases in Tizen.NUI.Samples.

this patch has requirement of dali-toolkit patch :
https://review.tizen.org/gerrit/#/c/platform/core/uifw/dali-toolkit/+/274611/

### API Changes ###

 - ACR:

### CODE REVIEW

**Comment by dongsug-song** [comment]
very nice for many useful samples!
but could you check the warnings as the below?

src/Tizen.NUI.Components/Controls/ScrollableBase.cs#L30
src/Tizen.NUI.Components/Controls/ScrollableBase.cs#L1804
...

these warnings are not related to this fix, but it would be better to resolve them when it comes.
(it could be shown in SVACE later on.)
thank you.

**Comment by everLEEst** [comment]
> very nice for many useful samples! but could you check the warnings as the below?
> 
> src/Tizen.NUI.Components/Controls/ScrollableBase.cs#L30 src/Tizen.NUI.Components/Controls/ScrollableBase.cs#L1804 ...
> 
> these warnings are not related to this fix, but it would be better to resolve them when it comes. (it could be shown in SVACE later on.) thank you.

no problem.
will update another PR.

**Comment by everLEEst** [comment]
https://review.tizen.org/gerrit/#/c/platform/core/uifw/dali-toolkit/+/274611/
is merged. patch is ready to review.

**Review by JoogabYun — APPROVED** [review]

File: src/Tizen.NUI.Components/Controls/ScrollableBase.cs [hunk 1]
```diff
@@ -1012,6 +1036,35 @@ public void ScrollToIndex(int index)
...
+                if (dist > 0)
+                {// if object is upper side
+                    targetPosition = scrollPosition - dist;
+                }
+                else
+                {// if object is down side
+                    targetPosition = scrollPosition - dist + childSize - viewSize;
+                }
+            }
+            ScrollTo(targetPosition, anim);
```
[the remark this answers is not in the document]
Comment (everLEEst) [reply src/Tizen.NUI.Components/Controls/ScrollableBase.cs:1065]: 음.. 우선 목표로 하는 건 저 함수를 나중에 key focus 뿐만 아니라 다른 포커스 상황에서도 항상 호출되게 하는 것을 가정하고 만든 함수입니다.
그래서 호출부에서 굳이 scroll 여부를 판단하지 않고 함수 내부에서 scroll이 필요한지를 판단해서 scroll 하는 것을 가정했습니다. 어차피 해당 함수에서 스크린의 위쪽이냐 아래쪽이냐에 따라 계산해야하는 값이 달라지기 때문에 geometry 값을 얻어오고 검사하는 과정은 필요합니다.

[the remark this answers is not in the document]
Comment (everLEEst) [reply src/Tizen.NUI.Components/Controls/ScrollableBase.cs]: 말씀해주신 내용이 좋을 거 같아 수정 반영 완료하였습니다 감사합니다.

**Comment by everLEEst** [comment]
> > very nice for many useful samples! but could you check the warnings as the below?
> > src/Tizen.NUI.Components/Controls/ScrollableBase.cs#L30 src/Tizen.NUI.Components/Controls/ScrollableBase.cs#L1804 ...
> > these warnings are not related to this fix, but it would be better to resolve them when it comes. (it could be shown in SVACE later on.) thank you.
> 
> no problem. will update another PR.



> > very nice for many useful samples! but could you check the warnings as the below?
> > src/Tizen.NUI.Components/Controls/ScrollableBase.cs#L30 src/Tizen.NUI.Components/Controls/ScrollableBase.cs#L1804 ...
> > these warnings are not related to this fix, but it would be better to resolve them when it comes. (it could be shown in SVACE later on.) thank you.
> 
> no problem. will update another PR.

https://github.com/Samsung/TizenFX/pull/4259#issuecomment-1125677401

**Review by JoogabYun — APPROVED** [review]

**Review by dongsug-song — APPROVED** [review]
approved

**Pull request merged**